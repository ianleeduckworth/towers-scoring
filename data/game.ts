export type Player = {
    got?: number;
    bet?: number;
}

export type Round = {
    roundNumber: number;
    player1: Player;
    player2: Player;
    player3?: Player;
    player4?: Player;
}

export type Game = {
    players: {
        player1: string;
        player2: string;
        player3?: string;
        player4?: string;
    },
    rounds: Round[];
}