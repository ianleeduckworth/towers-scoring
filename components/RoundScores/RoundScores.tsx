import * as React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import { Round, Players, Player } from '../../data/game';
import { defaults, globalStyles } from '../../styles/global';
import { calculateScore } from '../../utilities/scoreCalculator/scoreCalculator';

interface RoundScoresProps {
    round: Round;
    players: Players
}

export const RoundScores = ({ round, players }: RoundScoresProps) => {
    const getScore = (player: Player): number => {
        if (player.bet === undefined || player.got === undefined) {
            return 0;
        }

        return calculateScore(player.bet, player.got);
    }

    const getScoreStyling = (player?: Player) => {
        if (!player) {
            return styles.text;
        }

        const { bet, got } = player;
        if (bet === undefined || got === undefined) {
            return styles.text;
        }

        if (bet === got) {
            return styles.greenScoreText;
        }

        if (bet > got) {
            return styles.yellowScoreText;
        }

        if (got > bet) {
            return styles.redScoreText;
        }

        return styles.text;
    }

    return (
        <View style={styles.container}>
            <View style={styles.titleContainer}>
                <Text style={globalStyles.titleText}>Round {round.roundNumber}</Text>
            </View>
            <View style={styles.scoresContainer}>
                {/* Column 1 */}
                <View style={styles.playersColumn}>
                    <Text style={globalStyles.paragraphText}>{""}</Text>
                    <Text style={globalStyles.paragraphText}>{players.player1}</Text>
                    <Text style={globalStyles.paragraphText}>{players.player2}</Text>
                    {players.player3 && round.player3 && <Text style={globalStyles.paragraphText}>{players.player3}</Text>}
                    {players.player4 && round.player4 && <Text style={globalStyles.paragraphText}>{players.player4}</Text>}
                </View>
                {/* Column 2 */}
                <View style={styles.column}>
                    <Text style={globalStyles.paragraphText}>Bet</Text>
                    <Text style={styles.text}>{round.player1.bet}</Text>
                    <Text style={styles.text}>{round.player2.bet}</Text>
                    {players.player3 && round.player3 && <Text style={styles.text}>{round.player3.bet}</Text>}
                    {players.player4 && round.player4 && <Text style={styles.text}>{round.player4.bet}</Text>}
                </View>
                {/* Column 3 */}
                <View style={styles.column}>
                    <Text style={globalStyles.paragraphText}>Got</Text>
                    <Text style={styles.text}>{round.player1.got}</Text>
                    <Text style={styles.text}>{round.player2.got}</Text>
                    {players.player3 && round.player3 && <Text style={styles.text}>{round.player3.got}</Text>}
                    {players.player4 && round.player4 && <Text style={styles.text}>{round.player4.got}</Text>}
                </View>
                {/* Column 4 */}
                <View style={styles.column}>
                    <Text style={globalStyles.paragraphText}>Score</Text>
                    <Text style={getScoreStyling(round.player1)}>{getScore(round.player1)}</Text>
                    <Text style={getScoreStyling(round.player2)}>{getScore(round.player2)}</Text>
                    {players.player3 && round.player3 && <Text style={getScoreStyling(round.player3)}>{getScore(round.player3)}</Text>}
                    {players.player4 && round.player4 && <Text style={getScoreStyling(round.player4)}>{getScore(round.player4)}</Text>}
                </View>
            </View>

        </View>
    )
}

const styles = StyleSheet.create({
    container: {
        minWidth: '100%',
        height: 200,
        backgroundColor: defaults.lightGray,
        paddingLeft: 20,
        paddingRight: 20,
        borderWidth: 1,
        borderTopColor: defaults.black,
        borderBottomColor: defaults.black,
        marginTop: 10,
        marginBottom: 10,
    },
    titleContainer: {
        marginTop: 4,
        marginBottom: 4,
    },
    scoresContainer: {
        flexDirection: 'row',
    },
    playersColumn: {
        width: '40%'
    },
    column: {
        width: '20%',
    },
    text: {
        ...globalStyles.paragraphText,
        fontWeight: 'normal',
    },
    redScoreText: {
        ...globalStyles.paragraphText,
        fontWeight: 'normal',
        color: defaults.red
    },
    yellowScoreText: {
        ...globalStyles.paragraphText,
        fontWeight: 'normal',
        color: defaults.yellow
    },
    greenScoreText: {
        ...globalStyles.paragraphText,
        fontWeight: 'normal',
        color: defaults.green
    },
});