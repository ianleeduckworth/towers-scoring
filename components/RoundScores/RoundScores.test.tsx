import renderer from 'react-test-renderer';
import * as React from 'react';
import { RoundScores } from './RoundScores';
import { Game } from '../../data/game';

const baseGame: Game = {
    players: {
        player1: 'Player1',
        player2: 'Player2',
    },
    rounds: [{
        roundNumber: 1,
        player1: {
            bet: 20,
            got: 30,
        },
        player2: {
            bet: 20,
            got: 30,
        }
    }]
};

describe('round scores', () => {
    it('renders properly with 2 players', () => {
        const element = renderer.create(
            <RoundScores players={baseGame.players} round={baseGame.rounds[0]} />
        ).toJSON();

        expect(element).toMatchSnapshot();
    });

    it('renders properly with 3 players', () => {
        const game = {...baseGame};
        game.players.player3 = 'Player3';
        game.rounds[0].player3 = {
            bet: 20,
            got: 20,
        };

        const element = renderer.create(
            <RoundScores players={game.players} round={game.rounds[0]} />
        ).toJSON();

        expect(element).toMatchSnapshot();
    });

    it('renders properly with 4 players', () => {
        const game = {...baseGame};
        game.players.player3 = 'Player3';
        game.players.player4 = 'Player4';
        game.rounds[0].player3 = {
            bet: 20,
            got: 20,
        };
        game.rounds[0].player4 = {
            bet: 0,
            got: 0,
        };

        const element = renderer.create(
            <RoundScores players={game.players} round={game.rounds[0]} />
        ).toJSON();

        expect(element).toMatchSnapshot();
    });
})