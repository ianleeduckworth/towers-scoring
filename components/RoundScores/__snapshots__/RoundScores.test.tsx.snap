// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`round scores renders properly with 2 players 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F8F8F8",
      "borderBottomColor": "#000000",
      "borderTopColor": "#000000",
      "borderWidth": 1,
      "height": 200,
      "marginBottom": 10,
      "marginTop": 10,
      "minWidth": "100%",
      "paddingLeft": 20,
      "paddingRight": 20,
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 4,
        "marginTop": 4,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#000000",
          "fontFamily": "Roboto",
          "fontSize": 24,
          "fontWeight": "bold",
        }
      }
    >
      Round 
      1
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player1
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player2
      </Text>
    </View>
    <View
      style={
        Object {
          "width": "20%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Bet
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        20
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        20
      </Text>
    </View>
    <View
      style={
        Object {
          "width": "20%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Got
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        30
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        30
      </Text>
    </View>
    <View
      style={
        Object {
          "width": "20%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Score
      </Text>
      <Text
        style={
          Object {
            "color": "#B80000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        3
      </Text>
      <Text
        style={
          Object {
            "color": "#B80000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        3
      </Text>
    </View>
  </View>
</View>
`;

exports[`round scores renders properly with 3 players 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F8F8F8",
      "borderBottomColor": "#000000",
      "borderTopColor": "#000000",
      "borderWidth": 1,
      "height": 200,
      "marginBottom": 10,
      "marginTop": 10,
      "minWidth": "100%",
      "paddingLeft": 20,
      "paddingRight": 20,
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 4,
        "marginTop": 4,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#000000",
          "fontFamily": "Roboto",
          "fontSize": 24,
          "fontWeight": "bold",
        }
      }
    >
      Round 
      1
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player1
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player2
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player3
      </Text>
    </View>
    <View
      style={
        Object {
          "width": "20%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Bet
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        20
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        20
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        20
      </Text>
    </View>
    <View
      style={
        Object {
          "width": "20%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Got
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        30
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        30
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        20
      </Text>
    </View>
    <View
      style={
        Object {
          "width": "20%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Score
      </Text>
      <Text
        style={
          Object {
            "color": "#B80000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        3
      </Text>
      <Text
        style={
          Object {
            "color": "#B80000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        3
      </Text>
      <Text
        style={
          Object {
            "color": "#1AB800",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        32
      </Text>
    </View>
  </View>
</View>
`;

exports[`round scores renders properly with 4 players 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F8F8F8",
      "borderBottomColor": "#000000",
      "borderTopColor": "#000000",
      "borderWidth": 1,
      "height": 200,
      "marginBottom": 10,
      "marginTop": 10,
      "minWidth": "100%",
      "paddingLeft": 20,
      "paddingRight": 20,
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 4,
        "marginTop": 4,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#000000",
          "fontFamily": "Roboto",
          "fontSize": 24,
          "fontWeight": "bold",
        }
      }
    >
      Round 
      1
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player1
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player2
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player3
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Player4
      </Text>
    </View>
    <View
      style={
        Object {
          "width": "20%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Bet
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        20
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        20
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        20
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        0
      </Text>
    </View>
    <View
      style={
        Object {
          "width": "20%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Got
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        30
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        30
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        20
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        0
      </Text>
    </View>
    <View
      style={
        Object {
          "width": "20%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Score
      </Text>
      <Text
        style={
          Object {
            "color": "#B80000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        3
      </Text>
      <Text
        style={
          Object {
            "color": "#B80000",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        3
      </Text>
      <Text
        style={
          Object {
            "color": "#1AB800",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        32
      </Text>
      <Text
        style={
          Object {
            "color": "#1AB800",
            "fontFamily": "Roboto",
            "fontSize": 18,
            "fontWeight": "normal",
          }
        }
      >
        50
      </Text>
    </View>
  </View>
</View>
`;
